### ISSUE TITLE
Analyser: rearrange an equation AST for a given target variable

### ISSUE THREAD

**Issue by agarny** [cellml/libcellml#654]
Right now, we assume that a CellML equation always looks something like:

```
a = c+d-b
```

while CellML allows to have something like:

```
a+b = c+d
```

or even something like:

```
0 = c+d-b-a
```

So, after we get the AST for any of the above equation forms, we should be able to rearrange the AST so that we end up with an AST that corresponds to the first form.

We will therefore need to go through all the math operators and functions that CellML supports and make sure that we can handle them. For instance, if we have something like:

```
sin(x)+cos(y) = tan(z)
```

then it should be reformulated as something like:

```
x = asin(tan(z)-cos(y))
```

**Comment by kerimoyle** [comment]
Is this something that external libraries can do? I think SciPy does this kind of thing already ... is there something c++-ish?

**Comment by agarny** [comment]
Not SciPy, but [SymPy](https://www.sympy.org/) and its C++ equivalent is [SymEngine](https://github.com/symengine/symengine).

However, it looks like SymEngine still doesn't have support for MathML parsing (see https://github.com/symengine/symengine/issues/1062). So, we can't really use it as-is, but we could potentially write a MathML parser for what we need, if we really wanted to go that route.

This being said, we used to have an issue for using SymEngine (see issue #130), but @hsorby thought that we might as well write our own limited [CAS](https://en.wikipedia.org/wiki/Computer_algebra_system) (see https://github.com/cellml/libcellml/issues/130#issuecomment-413385082), hence we now have issue #232. So, I guess that this is in line with issue #232 and a subset of it.

**Comment by nickerso** [comment]
I'd suggest we close this one in favour of #232. This is not so much a subset of that, but more a use case for #232.

**Comment by agarny** [comment]
Well, my problem with #232 is that it's very generic, so not sure exactly what should be done about it. Here, at least, we have a very specific task.

**Comment by nickerso** [comment]
exactly, but to achieve this you need to write a minimal CAS...right? Just saying this task is not a separate piece of work and should be rolled into #232.
